File: java/state-of-tic-tac-toe/src/main/java/StateOfTicTacToe.java
import java.util.List;

class StateOfTicTacToe {

    private static final char X_PLAYER = 'X';
    private static final char O_PLAYER = 'O';

    public GameState determineState(String[] board) {
        String theBoard = String.join("", board);
        int numberOfXs = theBoard.replaceAll("[^X]", "").length();
        int numberOfOs = theBoard.replaceAll("[^O]", "").length();

        if ((numberOfXs - numberOfOs) > 1) {
            throw new IllegalArgumentException("Wrong turn order: X went twice");
        }

        if (numberOfOs > numberOfXs) {
            throw new IllegalArgumentException("Wrong turn order: O started");
        }

        boolean xWon = playerWon(X_PLAYER, theBoard);
        boolean oWon = playerWon(O_PLAYER, theBoard);

        if (xWon && (numberOfXs == numberOfOs)) {
            throw new IllegalArgumentException("Impossible board: game should have ended after the game was won");
        }

        if (xWon && oWon) {
            throw new IllegalArgumentException("Impossible board: game should have ended after the game was won");
        }

        if (xWon || oWon) {
            return GameState.WIN;
        }

        if (!theBoard.contains(" ")) {
            return GameState.DRAW;
        }

        return GameState.ONGOING;
    }

    private boolean playerWon(char player, String board) {

        List<List<Integer>> winningPositions = List.of(
                List.of(0, 1, 2), List.of(3, 4, 5), List.of(6, 7, 8), // horizontal
                List.of(0, 3, 6), List.of(1, 4, 7), List.of(2, 5, 8), // vertical
                List.of(0, 4, 8), List.of(2, 4, 6)                    // diagonal
        );

        return winningPositions.stream().anyMatch(pos -> wonAtPos(player, board, pos));
    }

    private boolean wonAtPos(char player, String board, List<Integer> position) {
        return position.stream().allMatch(i -> board.charAt(i) == player);
    }
}

File: javascript/state-of-tic-tac-toe/state-of-tic-tac-toe.js
export const gamestate = (board) => {

    checkBoard(board);

    return someoneWon(board) ? 'win' 
            : board.join('').includes(' ') ? 'ongoing' 
            : 'draw';
};

const someoneWon = (board) => {

    // Victory on rows
    if (board.includes('XXX') || board.includes('OOO')) {
        return true;
    }

    // Victory on columns
    for (let i = 0; i < 3; i++) {
        if (board[0][i] !== ' ' 
            && board [0][i] === board[1][i] && board[1][i] === board[2][i]) {
            return true;
        }
    }

    // Victory on a diagonal
    if (board[1][1] !== ' ' 
       && ((board[0][0] === board[1][1] && board[1][1] === board[2][2])
            || (board[0][2] === board[1][1] && board[1][1] === board[2][0]))) {
        return true;
    }

    return false;
};

const checkBoard = (board) => {
    let freq = {'X': 0, 'O': 0, ' ': 0};

    [...board.join('')].forEach(play => freq[play]++);

    if ((freq['X'] - freq['O']) > 1) {
        throw new Error('Wrong turn order: X went twice');
    }

    if (freq['O'] > freq['X']) {
        throw new Error('Wrong turn order: O started');
    }

    if (board.includes('XXX') && board.includes('OOO')) {
        throw new Error('Impossible board: game should have ended after the game was won');
    }
};
